### PR TITLE
Do not parenthesize exterior struct lit inside match guards

### PR DESCRIPTION
**Pull request by dtolnay**
Before this PR, the AST pretty-printer injects parentheses around expressions any time parens _could_ be needed depending on what else is in the code that surrounds that expression. But the pretty-printer did not pass around enough context to understand whether parentheses really _are_ needed on any particular expression. As a consequence, there are false positives where unneeded parentheses are being inserted.

Example:

```rust
#![feature(if_let_guard)]

macro_rules! pp {
    ($e:expr) => {
        stringify!($e)
    };
}

fn main() {
    println!("{}", pp!(match () { () if let _ = Struct {} => {} }));
}
```

**Before:**

```console
match () { () if let _ = (Struct {}) => {} }
```

**After:**

```console
match () { () if let _ = Struct {} => {} }
```

This PR introduces a bit of state that is passed across various expression printing methods to help understand accurately whether particular situations require parentheses injected by the pretty printer, and it fixes one such false positive involving match guards as shown above.

There are other parenthesization false positive cases not fixed by this PR. I intend to address these in follow-up PRs. For example here is one: the expression `{ let _ = match x {} + 1; }` is pretty-printed as `{ let _ = (match x {}) + 1; }` despite there being no reason for parentheses to appear there.

### CODE REVIEW

File: compiler/rustc_ast_pretty/src/pprust/state.rs [hunk 1]
```diff
@@ -1154,13 +1155,41 @@ impl<'a> State<'a> {
     }
 
     /// Print a `let pat = expr` expression.
-    fn print_let(&mut self, pat: &ast::Pat, expr: &ast::Expr) {
+    ///
+    /// Parentheses are inserted surrounding `expr` if a round-trip through the
+    /// parser would otherwise work out the wrong way in a condition position.
+    ///
+    /// For exammple each of the following would mean the wrong thing without
```
Comment (Veykril) [inline compiler/rustc_ast_pretty/src/pprust/state.rs:1162]:
```suggestion
    /// For example each of the following would mean the wrong thing without
```

**Comment by dtolnay** [comment]
- Rebased to resolve conflict with https://github.com/rust-lang/rust/pull/118527

**Review by compiler-errors — APPROVED** [review]
This is great, thanks

**Pull request merged**